x-pack/plugins/maps/public/components/map/mb/view.js

```diff
@@ -6,154 +6,252 @@
-        return;
-      }
-
-      const eventType = args[0];
-      const mbLayerId = args[1];
-      const handler = args[2];
-      this._addListener(eventType, mbLayerId, handler);
-
-      originalMbBoxOnFunc.apply(this._mbMap, args);
-    };
-    this._mbMap.removeLayer = (id) => {
-      this._removeListeners(id);
-      originalMbBoxRemoveLayerFunc.apply(this._mbMap, [id]);
-    };
-
-    this._initResizerChecker();
-
-    // moveend callback is debounced to avoid updating map extent state while map extent is still changing
-    // moveend is fired while the map extent is still changing in the following scenarios
-    // 1) During opening/closing of layer details panel, the EUI animation results in 8 moveend events
-    // 2) Setting map zoom and center from goto is done in 2 API calls, resulting in 2 moveend events
-    this._mbMap.on('moveend', _.debounce(() => {
-      this.props.extentChanged(this._getMapState());
-    }, 100));
-
-    const throttledSetMouseCoordinates = _.throttle(e => {
-      this.props.setMouseCoordinates({
-        lat: e.lngLat.lat,
-        lon: e.lngLat.lng
-      });
-    }, 100);
-    this._mbMap.on('mousemove', throttledSetMouseCoordinates);
-    this._mbMap.on('mouseout', () => {
-      throttledSetMouseCoordinates.cancel(); // cancel any delayed setMouseCoordinates invocations
-      this.props.clearMouseCoordinates();
-    });
-
-    this.props.onMapReady(this._getMapState());
-  }
-
-  _addListener(eventType, mbLayerId, handler) {
-    this._removeListener(eventType, mbLayerId);
-
-    const eventHandlers = !this._listeners.has(mbLayerId)
-      ? new Map()
-      : this._listeners.get(mbLayerId);
-    eventHandlers.set(eventType, handler);
-    this._listeners.set(mbLayerId, eventHandlers);
-  }
-
-  _removeListeners(mbLayerId) {
-    if (this._listeners.has(mbLayerId)) {
-      const eventHandlers = this._listeners.get(mbLayerId);
-      eventHandlers.forEach((value, eventType) => {
-        this._removeListener(eventType, mbLayerId);
-      });
-      this._listeners.delete(mbLayerId);
-    }
-  }
-
-  _removeListener(eventType, mbLayerId) {
-    if (this._listeners.has(mbLayerId)) {
-      const eventHandlers = this._listeners.get(mbLayerId);
-      if (eventHandlers.has(eventType)) {
-        this._mbMap.off(eventType, mbLayerId, eventHandlers.get(eventType));
-        eventHandlers.delete(eventType);
-      }
-    }
-  }
-
-  _initResizerChecker() {
-    this._checker = new ResizeChecker(this.refs.mapContainer);
-    this._checker.on('resize', () => {
-      this._mbMap.resize();
-    });
-  }
+   _onMouseMove = _.debounce(async (e) => {
+
+     const mbLayerIds = this._getMbLayerIdsForTooltips();
+     const features = this._mbMap.queryRenderedFeatures(e.point, { layers: mbLayerIds });
+
+     if (!features.length) {
+       this.props.setTooltipState(null);
+       return;
+     }
+
+     const targetFeature = features[0];
+     if (this.props.tooltipState) {
```

Comment:
Do we need this check. Should we update the tooltip lat/lon as the user continues to move inside of large features?

Reply:
I wouldn't. it's not doing that now.

It's getting to video-game-ish when it does move with pointer imho. When we do the on-click handling, we might want to put the tooltip at the place of click/center of feature,...